File: dungeons-26.1.2/src/main/java/fishmod/features/other/WardrobeHotkeys.java
package fishmod.features.other;

import fishmod.utils.Keybinds;
import fishmod.utils.config.values.FishSettings;
import net.fabricmc.fabric.api.client.event.lifecycle.v1.ClientTickEvents;
import net.minecraft.client.KeyMapping;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.screens.inventory.AbstractContainerScreen;
import net.minecraft.client.input.KeyEvent;
import net.minecraft.client.input.MouseButtonEvent;
import net.minecraft.core.registries.BuiltInRegistries;
import net.minecraft.world.inventory.AbstractContainerMenu;
import net.minecraft.world.inventory.ContainerInput;
import net.minecraft.world.inventory.Slot;
import net.minecraft.world.item.ItemStack;

/**
 * Wardrobe/Loadouts quick-swap: pressing FishMod's slot-N hotkey (keyboard or mouse button,
 * whatever it's bound to in Controls) clicks the matching slot in the currently open Wardrobe
 * (Armor Sets) or Loadouts GUI.
 *
 * Loadouts uses a fixed slot layout (verified in-game): 3 columns x 4 rows of "select this
 * loadout" icons at raw slot indices 14/15/16, 23/24/25, 32/33/34, 41/42/43 — hardcoded below.
 *
 * Wardrobe's clickable "select this set" icon moves depending on how many sets are on the page,
 * so instead of a fixed index it's found each time by scanning the hotkey's column for
 * Hypixel's wool/dye/barrier icon.
 *
 * The actual click is deferred by one client tick after the key/click event, since firing it
 * synchronously in the same tick as the input event was causing visual glitches in Hypixel's GUI.
 */
public class WardrobeHotkeys {

    private static final int PLAYER_INV_SLOTS = 36;

    /** Raw slot index for Loadout hotkeys 1-12, in the same row-major order as Keybinds.wardrobeSlots. */
    private static final int[] LOADOUT_SLOTS = {14, 15, 16, 23, 24, 25, 32, 33, 34, 41, 42, 43};

    private static Runnable pendingClick;
    private static int pendingTicks;

    public static void init() {
        ClientTickEvents.END_CLIENT_TICK.register(client -> {
            if (pendingClick == null) return;
            if (pendingTicks > 0) {
                pendingTicks--;
                return;
            }
            Runnable action = pendingClick;
            pendingClick = null;
            action.run();
        });
    }

    public static boolean keyPressed(KeyEvent input, AbstractContainerScreen<?> screen) {
        return tryActivate(screen, mapping -> mapping.matches(input));
    }

    public static boolean mouseClicked(MouseButtonEvent click, AbstractContainerScreen<?> screen) {
        return tryActivate(screen, mapping -> mapping.matchesMouse(click));
    }

    private static boolean tryActivate(AbstractContainerScreen<?> screen, java.util.function.Predicate<KeyMapping> matches) {
        if (!FishSettings.wardrobeHotkeysEnabled) return false;

        String title = screen.getTitle().getString().replaceAll("§.", "").trim();
        boolean isWardrobe = title.contains("Armor Sets") || title.equals("Wardrobe");
        boolean isLoadout = title.contains("Loadouts");
        if (!isWardrobe && !isLoadout) return false;

        AbstractContainerMenu handler = screen.getMenu();
        int containerSize = handler.slots.size() - PLAYER_INV_SLOTS;
        // Sanity check: this must actually be a chest-style GUI, not some other screen
        // that happens to share a title substring.
        if (containerSize < 27 || containerSize % 9 != 0) return false;

        for (int i = 0; i < Keybinds.wardrobeSlots.length; i++) {
            KeyMapping mapping = Keybinds.wardrobeSlots[i];
            if (mapping == null || mapping.isUnbound() || !matches.test(mapping)) continue;

            Slot target = resolveTarget(handler, containerSize, isWardrobe, i);
            if (target == null) return false;

            int containerId = handler.containerId;
            int slotId = target.index;
            pendingClick = () -> {
                Minecraft mc = Minecraft.getInstance();
                if (mc.player == null || mc.gameMode == null) return;
                // handleContainerInput(containerId, slotId, button, ContainerInput, player) — button
                // 0 = left click, matching the old clickSlot(syncId, slotId, button, actionType, player).
                mc.gameMode.handleContainerInput(containerId, slotId, 0, ContainerInput.PICKUP, mc.player);
                // screen.onClose() (not player.closeContainer()) — matches what pressing Escape
                // does: sends the close packet AND actually dismisses the on-screen GUI. Calling
                // just closeContainer() left the GUI widget on screen out of sync with the
                // now-reset player.containerMenu, which showed up as a close/reopen/close
                // flicker once the server's own state caught up.
                if (FishSettings.wardrobeHotkeysAutoClose && mc.screen == screen) {
                    screen.onClose();
                }
            };
            pendingTicks = 1;
            return true;
        }

        return false;
    }

    private static Slot resolveTarget(AbstractContainerMenu handler, int containerSize, boolean isWardrobe, int hotkeyIndex) {
        if (isWardrobe) {
            return findSelectSlot(handler, containerSize, hotkeyIndex);
        }
        if (hotkeyIndex >= LOADOUT_SLOTS.length) return null;
        int slotIndex = LOADOUT_SLOTS[hotkeyIndex];
        return slotIndex < handler.slots.size() ? handler.slots.get(slotIndex) : null;
    }

    /** Scans column {@code column} (of a 9-wide grid) for Hypixel's wool/dye/barrier "select" icon. */
    private static Slot findSelectSlot(AbstractContainerMenu handler, int containerSize, int column) {
        if (column >= 9) return null;
        for (int i = column; i < containerSize; i += 9) {
            Slot slot = handler.slots.get(i);
            ItemStack stack = slot.getItem();
            if (stack.isEmpty()) continue;

            String id = vanillaId(stack);
            if (id == null) continue;

            if (id.endsWith("_wool") || id.endsWith("_dye") || id.endsWith("_stained_glass_pane") || id.equals("minecraft:barrier")) {
                return slot;
            }
        }
        return null;
    }

    private static String vanillaId(ItemStack stack) {
        try {
            return BuiltInRegistries.ITEM.getKey(stack.getItem()).toString();
        } catch (Exception e) {
            return null;
        }
    }
}

File: dungeons-26.1.2/src/main/kotlin/fishmod/features/other/WardrobeHotkeys.kt
package fishmod.features.other

import fishmod.utils.Keybinds
import fishmod.utils.config.values.FishSettings
import net.fabricmc.fabric.api.client.event.lifecycle.v1.ClientTickEvents
import net.minecraft.client.KeyMapping
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.screens.inventory.AbstractContainerScreen
import net.minecraft.client.input.KeyEvent
import net.minecraft.client.input.MouseButtonEvent
import net.minecraft.core.registries.BuiltInRegistries
import net.minecraft.world.inventory.AbstractContainerMenu
import net.minecraft.world.inventory.ContainerInput
import net.minecraft.world.inventory.Slot
import net.minecraft.world.item.ItemStack
import java.util.function.Predicate

object WardrobeHotkeys {

    private const val PLAYER_INV_SLOTS = 36

    private val COLOR = fishmod.utils.Constants.STRIP_COLOR_REGEX

    private val LOADOUT_SLOTS = intArrayOf(14, 15, 16, 23, 24, 25, 32, 33, 34, 41, 42, 43)

    private var pendingClick: Runnable? = null
    private var pendingTicks = 0

    @JvmStatic
    fun init() {
        ClientTickEvents.END_CLIENT_TICK.register { _ ->
            val click = pendingClick
            if (click == null) return@register
            if (pendingTicks > 0) {
                pendingTicks--
                return@register
            }
            pendingClick = null
            click.run()
        }
    }

    @JvmStatic
    fun keyPressed(input: KeyEvent, screen: AbstractContainerScreen<*>): Boolean {
        return tryActivate(screen) { mapping -> mapping.matches(input) }
    }

    @JvmStatic
    fun mouseClicked(click: MouseButtonEvent, screen: AbstractContainerScreen<*>): Boolean {
        if (click.button() == 0 || click.button() == 1) {
            val hovered = (screen as fishmod.mixin.accessors.HandledScreenAccessor).`fishmod$getHoveredSlot`()
            if (hovered != null && !hovered.item.isEmpty) return false
        }
        return tryActivate(screen) { mapping -> mapping.matchesMouse(click) }
    }

    private fun tryActivate(screen: AbstractContainerScreen<*>, matches: Predicate<KeyMapping>): Boolean {
        if (!FishSettings.wardrobeHotkeysEnabled) return false
        val slots = Keybinds.wardrobeSlots ?: return false

        val title = fishmod.utils.ScreenTitle.plain(screen).trim()
        val isWardrobe = title.contains("Armor Sets") || title == "Wardrobe"
        val isLoadout = title.contains("Loadouts")
        if (!isWardrobe && !isLoadout) return false

        val handler: AbstractContainerMenu = screen.menu
        val containerSize = handler.slots.size - PLAYER_INV_SLOTS
        if (containerSize < 27 || containerSize % 9 != 0) return false

        if (tryPageTurn(handler, containerSize, screen, Keybinds.wardrobeNextPage, matches, "next page")) return true
        if (tryPageTurn(handler, containerSize, screen, Keybinds.wardrobePrevPage, matches, "previous page")) return true

        for (i in slots.indices) {
            val mapping = slots[i]
            if (mapping == null || mapping.isUnbound || !matches.test(mapping)) continue

            val target = resolveTarget(handler, containerSize, isWardrobe, i) ?: return false

            val containerId = handler.containerId
            val slotId = target.index
            pendingClick = Runnable {
                val mc = Minecraft.getInstance()
                val mcPlayer = mc.player
                if (mcPlayer == null || mc.gameMode == null) return@Runnable
                mc.gameMode!!.handleContainerInput(containerId, slotId, 0, ContainerInput.PICKUP, mcPlayer)
                if (FishSettings.wardrobeHotkeysAutoClose && mc.screen === screen) {
                    screen.onClose()
                }
            }
            pendingTicks = 1
            return true
        }

        return false
    }

    private fun tryPageTurn(handler: AbstractContainerMenu, containerSize: Int, screen: AbstractContainerScreen<*>, mapping: KeyMapping?, matches: Predicate<KeyMapping>, label: String): Boolean {
        if (mapping == null || mapping.isUnbound || !matches.test(mapping)) return false

        val target = findByName(handler, containerSize, label) ?: return false

        val containerId = handler.containerId
        val slotId = target.index
        pendingClick = Runnable {
            val mc = Minecraft.getInstance()
            val mcPlayer = mc.player
            if (mcPlayer == null || mc.gameMode == null) return@Runnable
            mc.gameMode!!.handleContainerInput(containerId, slotId, 0, ContainerInput.PICKUP, mcPlayer)
        }
        pendingTicks = 1
        return true
    }

    private fun findByName(handler: AbstractContainerMenu, containerSize: Int, label: String): Slot? {
        for (i in 0 until containerSize) {
            val slot = handler.slots[i]
            val stack = slot.item
            if (stack.isEmpty) continue
            val name = stack.hoverName.string.replace(COLOR, "").trim()
            if (name.lowercase().contains(label)) return slot
        }
        return null
    }

    private fun resolveTarget(handler: AbstractContainerMenu, containerSize: Int, isWardrobe: Boolean, hotkeyIndex: Int): Slot? {
        if (isWardrobe) {
            return findSelectSlot(handler, containerSize, hotkeyIndex)
        }
        if (hotkeyIndex >= LOADOUT_SLOTS.size) return null
        val slotIndex = LOADOUT_SLOTS[hotkeyIndex]
        return if (slotIndex < handler.slots.size) handler.slots[slotIndex] else null
    }

    private fun findSelectSlot(handler: AbstractContainerMenu, containerSize: Int, column: Int): Slot? {
        if (column >= 9) return null
        var i = column
        while (i < containerSize) {
            val slot = handler.slots[i]
            val stack: ItemStack = slot.item
            if (!stack.isEmpty) {
                val id = vanillaId(stack)
                if (id != null &&
                    (id.endsWith("_wool") || id.endsWith("_dye") || id.endsWith("_stained_glass_pane") || id == "minecraft:barrier")
                ) {
                    return slot
                }
            }
            i += 9
        }
        return null
    }

    private fun vanillaId(stack: ItemStack): String? {
        return try {
            BuiltInRegistries.ITEM.getKey(stack.item).toString()
        } catch (e: Exception) {
            null
        }
    }
}
